azurerm/resource_arm_application_gateway.go

```diff
@@ -0,0 +1,1691 @@
+			"frontend_port_name":             path.Base(*config.ApplicationGatewayHTTPListenerPropertiesFormat.FrontendPort.ID),
+			"frontend_port_id":               *config.ApplicationGatewayHTTPListenerPropertiesFormat.FrontendPort.ID,
+			"protocol":                       string(config.ApplicationGatewayHTTPListenerPropertiesFormat.Protocol),
+		}
+
+		if config.ApplicationGatewayHTTPListenerPropertiesFormat.HostName != nil {
+			listener["host_name"] = *config.ApplicationGatewayHTTPListenerPropertiesFormat.HostName
+		}
+
+		if config.ApplicationGatewayHTTPListenerPropertiesFormat.SslCertificate != nil {
+			listener["ssl_certificate_name"] = path.Base(*config.ApplicationGatewayHTTPListenerPropertiesFormat.SslCertificate.ID)
+			listener["ssl_certificate_id"] = *config.ApplicationGatewayHTTPListenerPropertiesFormat.SslCertificate.ID
+		}
+
+		if config.ApplicationGatewayHTTPListenerPropertiesFormat.RequireServerNameIndication != nil {
+			listener["require_sni"] = *config.ApplicationGatewayHTTPListenerPropertiesFormat.RequireServerNameIndication
+		}
+
+		result = append(result, listener)
+	}
+
+	return result
+}
+
+func flattenApplicationGatewayProbes(probes *[]network.ApplicationGatewayProbe) []interface{} {
+	result := make([]interface{}, 0, len(*probes))
+
+	for _, config := range *probes {
+		settings := map[string]interface{}{
+			"id":                  *config.ID,
+			"name":                *config.Name,
+			"protocol":            string(config.ApplicationGatewayProbePropertiesFormat.Protocol),
+			"path":                *config.ApplicationGatewayProbePropertiesFormat.Path,
+			"host":                *config.ApplicationGatewayProbePropertiesFormat.Host,
+			"interval":            int(*config.ApplicationGatewayProbePropertiesFormat.Interval),
+			"timeout":             int(*config.ApplicationGatewayProbePropertiesFormat.Timeout),
+			"unhealthy_threshold": int(*config.ApplicationGatewayProbePropertiesFormat.UnhealthyThreshold),
+		}
+
+		result = append(result, settings)
+	}
+
+	return result
+}
+
+func flattenApplicationGatewayRequestRoutingRules(rules *[]network.ApplicationGatewayRequestRoutingRule) []interface{} {
+	result := make([]interface{}, 0, len(*rules))
+
+	for _, config := range *rules {
+		listener := map[string]interface{}{
+			"id":                 *config.ID,
+			"name":               *config.Name,
+			"rule_type":          string(config.ApplicationGatewayRequestRoutingRulePropertiesFormat.RuleType),
+			"http_listener_id":   *config.ApplicationGatewayRequestRoutingRulePropertiesFormat.HTTPListener.ID,
+			"http_listener_name": path.Base(*config.ApplicationGatewayRequestRoutingRulePropertiesFormat.HTTPListener.ID),
+		}
+
+		if config.ApplicationGatewayRequestRoutingRulePropertiesFormat.BackendAddressPool != nil {
+			listener["backend_address_pool_name"] = path.Base(*config.ApplicationGatewayRequestRoutingRulePropertiesFormat.BackendAddressPool.ID)
```

Comment:
we're assuming here the name will be at the end - can we switch this to using the `parseAzureResourceID` function to parse the values out of the URL to guarantee that this is the case?